linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -5140,6 +5205,41 @@ private[optimizer] object OptimizerCore {
     }
   }
 
+  final class InlineableFieldBodies(
+    val fieldBodies: Map[(ClassName, FieldName), InlineableFieldBodies.FieldBody]
+  ) {
+    override def equals(that: Any): Boolean = that match {
+      case that: InlineableFieldBodies =>
+        this.fieldBodies == that.fieldBodies
+      case _ =>
+        false
+    }
+
+    override def hashCode(): Int = fieldBodies.##
+
+    override def toString(): String = {
+      fieldBodies
+        .map(f => s"${f._1._1.nameString}::${f._1._2.nameString}: ${f._2}")
+        .mkString("InlineableFieldBodies(", ", ", ")")
+    }
+  }
+
+  object InlineableFieldBodies {
+    sealed abstract class FieldBody
```

Comment:
I do not understand why we need this additional class structure. Aren't these the same as trees?

Reply:
The idea here was to have something that we can reliably compare for equality, for the incremental invalidation. Normally we don't compare general Trees for equality.

Reply:
I see. IIUC, comparing trees for equality in general is safe, right? It's just expensive.

So having this additional structure doesn't seem to actually buy us anything in that regard; or am I missing something?